dhall-json/src/Dhall/JSONToDhall.hs

```diff
@@ -1103,6 +1109,93 @@ dhallFromJSON (Conversion {..}) expressionType =
+      else go xs
+    go (_:xs) = go xs
+    isSchema recordItems
+      -- TODO: check that `Type` is a Record Type and `default` is a Record literal
+      | Map.keysSet recordItems == Set.fromList [Text.pack "Type", Text.pack "default"] = True
+      | otherwise = False
+    getCompletableRecord :: Text -> Map.Map Text ExprX -> (Text, CompletableRecord)
+    getCompletableRecord name fields = (name, (go' "Type", go' "default"))
+      where
+        go' attr = case Map.lookup attr fields of
+          Just (D.RecordLit schema) -> Map.toList schema
+          Just (D.Record schema) -> Map.toList schema
+          _ -> []
+getCompletableRecords _ = []
+
+type ExprS = Expr Src D.Import
+
+-- Given a record attributes map, look for matching completable record and return the minimum attributes that complete the record
+findMatchingRecords :: Map.Map Text ExprS -> [(Text, CompletableRecord)] -> [(Text, Map.Map Text ExprS)]
+findMatchingRecords schema = map go
+  where
+    go :: (Text, CompletableRecord) -> (Text, Map.Map Text ExprS)
+    go (name, (rtype, rdefault)) = -- trace ("Testing " <> Text.unpack name <> " : " <> show (Map.fromList rtype)) $
+      if recordFit (Map.fromList rtype) rdefault
+      then (name, removeDefault rdefault)
+      else (name, Map.empty)
+
+    -- TODO: check for default and matching types
+    recordFit rtype rdefault
+      | Map.keysSet schema `Set.isSubsetOf` Map.keysSet rtype = True
+      | otherwise = False
+
+    -- TODO: remove un-necessary defaults
+    removeDefault _ = schema
+
+dhallFromJSONSchemas :: FilePath -> ExprX -> ExprS -> Either CompileError ExprS
+dhallFromJSONSchemas filePath schemas (D.RecordLit schema) =
+  do
+    if null completableRecords
+      then Left NoSchemas
+      else if null completedRecords
+           then Left NoMatchingSchema
+           else Right $ mkExpression $ head completedRecords
+  where
+    completableRecords = getCompletableRecords schemas
+    completedRecords :: [(Text, Map.Map Text ExprS)]
+    completedRecords =
+      filter
+        (\(_, attr) -> attr /= Map.empty)
+        (findMatchingRecords schema completableRecords)
+    mkExpression :: (Text, Map.Map Text ExprS) -> ExprS
+    mkExpression (name, attrs) = D.Let
+      (D.Binding {
+          bindingSrc0 = Nothing,
+          variable = "schemas",
+          bindingSrc1 = Nothing,
+          annotation = Nothing,
+          bindingSrc2 = Nothing,
+          value = D.Embed mkImport
+          })
+      (mkRecordCompletion name attrs)
+    mkRecordCompletion :: Text -> Map.Map Text ExprS -> ExprS
+    mkRecordCompletion name attrs = D.RecordCompletion (D.Field (D.Var (D.V "schemas" 0)) name) (D.RecordLit attrs)
+    mkImport = D.Import {
+      importMode = D.Code,
+      importHashed = D.ImportHashed {
+          hash = Nothing,
+          importType = D.Local D.Here (D.File {directory = D.Directory {components = []},
+                                               file = Text.pack (drop 2 filePath) })
```

Comment:
TODO: make this work for path that are not local relative (e.g. not `./file.dhall`)